File: priorityqueue.cpp
#include <iostream>
#include <limits>
using namespace std;
void swap(int *x, int *y)
{
    int temp = *x;
    *x = *y;
    *y = temp;
}
class maxheap{
  public:
    int *heap; // pointer to array of elements in heap
    int maxsize; // maximum possible size of min heap
    int size; // Current number of elements in min heap
    maxheap(int data);
    int parent(int i){return (i-1)/2;}
    int leftchild(int i){return 2*i +1;}
    int rightchild(int i){return 2*i +2;}
    void insert(int i);
    void siftup(int i);
    void siftdown(int i);
    int extractmax();
    void remove(int i);
};
maxheap::maxheap(int data){
      maxsize = data;
      heap = new int[maxsize];
      size = -1;
}
void maxheap::siftup(int i){
  while(i > -1 && heap[parent(i)] < heap[i]){
    swap(&heap[parent(i)],&heap[i]);
    i = parent(i);
    }
}
void maxheap::insert(int i){
      if(size == maxsize){return;}
      size += 1;
      heap[size] = i;
      maxheap::siftup(size);
}
void maxheap::siftdown(int i){
    int maxindex = i;
    int l = leftchild(i);
    if(l < size && heap[l] > heap[i]){
      maxindex = l;
    }
    int r = rightchild(i);
    if(r < size && heap[r] > heap[i]){
      maxindex = r;
    }
    if(i != maxindex){
      swap(&heap[maxindex],&heap[i]);
      maxheap::siftdown(maxindex);
    }
}
int maxheap::extractmax(){
    int result = heap[0];
    heap[0] = heap[size];
    size = size-1;
    maxheap::siftdown(0);
    return result;
}
void maxheap::remove(int i){
  heap[i] = std::numeric_limits<int>::max();
  maxheap::siftup(i);
  maxheap::extractmax();
}
int main() {
  maxheap h(20);
  h.insert(2);
  h.insert(3);
  h.insert(4);
  h.insert(5);
  h.insert(1);
  h.insert(12);
  h.insert(22);
  h.insert(47);
  h.insert(9);
  h.insert(9);
  h.insert(17);
  h.insert(109);
  h.insert(32);
  for(int i = 0; i < h.size; i++){
    cout << h.heap[i] << endl;
  }

  

}

File: priorityqueue.py
class maxheap:
  def __init__(self):
      #self.maxsize = data
      self.heap = []
      self.size = -1
  def parent(self,i):
    return(int((i-1)/2))
  def leftchild(self,i):
    return 2*i + 1
  def rightchild(self,i):
    return 2*i + 2
  def siftup(self,i):
    while (i > -1) and (self.heap[i] > self.heap[self.parent(i)]):
      self.heap[i],self.heap[self.parent(i)] = self.heap[self.parent(i)],self.heap[i]
      i = self.parent(i)
  def siftdown(self,i):
    maxindex = i
    l = self.leftchild(i)
    if(l < self.size and self.heap[l] > self.heap[maxindex]):
      maxindex = l
    r = self.rightchild(i)
    if(r < self.size and self.heap[r] > self.heap[maxindex]):
      maxindex = r
    if (i != maxindex ):
      self.heap[i],self.heap[maxindex] = self.heap[maxindex],self.heap[i]
      self.siftdown(maxindex)
  def extractmax(self):
    result = self.heap[0]
    self.heap[0] = self.heap[self.size]
    self.size = self.size -1
    self.siftdown(0) 
    return result
  def insert(self,p):
    #if (self.size == self.maxsize):
      #print("Error")
      #return
    self.size += 1
    self.heap.append(p)
    self.siftup(self.size)
  def remove(self,i):
    self.heap[i] = (float('inf'))
    self.siftup(i)
    self.extractmax()


h = maxheap()
h.insert(2)
h.insert(3)
h.insert(4)
h.insert(5)
h.insert(1)
h.insert(12)
h.insert(22)
h.insert(47)
h.insert(9)
h.insert(9)
h.insert(17)
h.insert(109)
h.insert(32)
print(h.heap)
while(h.size >= 0 ):
  print(h.extractmax())





    

